Encryption.md

```diff
@@ -75,55 +75,126 @@ written to the output stream.
-written after each ciphertext. The IVs (12 bytes) are written before each ciphertext.
+All modules are encrypted by an AES-GCM cipher, without padding. Unique IVs are generated and 
+written before each ciphertext. The IV length is 12 
+bytes (96 bits), which is generally considered to be the optimal size of GCM IVs.  The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” to be empty for 
+this IV length (see section 8.2.2). 
+
+The authentication tags (16 bytes) are written after each ciphertext.
 
 ### AES_GCM_CTR_V1
-Thrift modules are encrypted with the AES-GCM cipher, as described above. 
-The pages are encrypted with AES-CTR, where the IVs (16 bytes) are written before each 
-ciphertext.
+Thrift modules are encrypted with AES-GCM, as described above. The pages are encrypted by an AES-CTR
+cipher without padding. The nonces are generated and written before each ciphertext, with the nonce length of 16 bytes 
+(128 bits). At least 12 bytes out of 16 are constructed with an RBG.
+
+```c
+struct AesGcmV1 {
+  /** Retrieval metadata of AAD prefix used for encryption of pages and structures **/
+  1: optional binary aad_metadata
+}
+
+struct AesGcmCtrV1 {
+  /** Retrieval metadata of AAD prefix used for encryption of structures **/
+  1: optional binary aad_metadata
+}
+
+union EncryptionAlgorithm {
+  1: AesGcmV1 AES_GCM_V1
+  2: AesGcmCtrV1 AES_GCM_CTR_V1
+}
+```
+
+### Additional Authenticated Data
+
+The AES GCM cipher in the basic mode protects against byte replacement inside a ciphertext, 
+but can't prevent replacement of one ciphertext with another (encrypted with the same key). 
+This can be solved by the AAD mode, using different AADs for different ciphertexts. Parquet 
+modular encryption uses the AAD mode to protect against swapping ciphertext modules inside a 
+file, between files - or against swapping full files (for example, replacement of a file with an 
+old version). Obviously, the AAD must reflect the identity of a file and of the modules inside 
+the file. 
+
+Parquet constructs a module AAD from two components: an optional AAD prefix - a string provided 
+by the user for the file, and an AAD suffix, built internally for each encrypted module inside 
+the file. The AAD prefix should reflect the target identity that helps to detect file swapping - 
+for example, table name and version or date / timestamp (e.g., "employees_23_05_2018"). The 
+AAD suffix reflects the internal identity of modules inside the file, which for example prevents 
+replacement of column pages in row group 0 by pages from the same column in row group 1. The module 
+AAD is a direct concatenation of the prefix and suffix parts. 
+
+#### AAD prefix
+
+A file writer passes an AAD prefix string upon file creation, that allows to differentate it from other
+files and datasets. The reader should know this string apriori, or should be able to retrieve and verify
+it, using a convention accepted in the organization.
```

Comment:
Why not generate a random AAD prefix?

If Parquet relies on users to supply a prefix, then it allows the case where files are not protected by an AAD or where files use the same AAD. That allows attackers to swap ciphertext streams when the user has been lazy or doesn't understand the purpose of the AAD.

A random AAD prefix would ensure that ciphertext cannot be moved between files and the AAD suffix ensures that ciphertext cannot be moved between row groups or columns.

Reply:
Another alternative is to allow users to pass a prefix, but to add a random portion to the suffix that makes the file AADs unique.

Reply:
Random AAD prefix is possible, but requires a trusted service where these prefixes are stored for every file. A reader must be able to retrieve an AAD prefix for every file. Since the data storage is untrusted (thats why we use the keys and AADs in the first place), the AAD prefixes must be stored in a secure location, and serviced by an authenticated process.
Unlike keys, AADs are not secret. Hence, having a "well known" method of AAD prefix creation (such as \<table name\>\_\<date\> or \<table name\>\_\<version\>) allows writers and readers to create/verify the file AAD prefix without a remote service.
Still, I'll add your suggestion as an example of possible AAD prefixes and their store/fetch service.

Reply:
AAD prefix is not necessarily stored in the aad_metadata. The "well known" method approach, described above, doesn't require setting the aad_metadata at all. But I agree it does require smart users, that can generate and verify unique AAD prefixes. Dumb users should work with a trusted service that keeps the random AAD prefixes created for each file. I'll add this to the text (maybe avoiding the dumb/smart terminology :). Similar to key metadata section, we'll have a number of example usages of the aad metadata field.

Reply:
btw, the random AAD prefix won't be stored in aad_metadata - no point, because aad_metadata is unprotected/untrusted, and a reader doesn't have a way to verify it other than to go to an authenticated service and get a correct AAD prefix for this file.

Reply:
Thinking of this, there is an additional option: get a descriptive prefix from the user, add a random part to it to make it truly unique (as you suggested) - and send it to a trusted service to hash/sign it with an asymmetric private key. The result is sent back to the file writer, and stored inside the aad_metadata field. The file reader will have the certificate with  the public key, that can be used to very the AAD prefix stored in the file. The reader will use the "descriptive" part to make sure it is the file it needs.
Its a kind of an intermediate option - it still demands a trusted service, but no need in storing all AADs there for each file, the service just signs the AADs and sends them back. It also still requires some smarts on the user part (both writer and reader), to create a good "descriptive prefix".

Reply:
A problem with this intermediate approach - say somebody (maliciously or not) creates a new file with the same descriptive AAD prefix part (and a different random part), and gets it signed. Then the original file can be replaced by this one, and it will pass the reader checks.
So we're back to the two main options: smart users with descriptive unique AAD prefixes and no service overhead; or dumb users with random AAD prefixes and a trusted storage service.

Reply:
That sounds complicated.

What we really need is to ensure that there is some unique random string for each file, and to allow the user to add an additional AAD string. If the user can add to the AAD, then there is no need to hide the unique random. Anything that requires a secret AAD portion can be done in the user section.

How about this: let's leave `aad_metadata` as it is for user metadata and add a per-file unique to the `EncryptionAlgorithm` structs, called `aad_file_unique`. Then each AES-GCM stream would concatenate `aad_metadata`, `aad_file_unique`, and the AAD suffix to form each AAD.

That ensures that blocks can't be swapped between files and allows the user to have custom AAD content that is stored in the file or not.

Reply:
I think there's also confusion about whether the per-file random can be public or not. It can be. The attack that it prevents is swapping data encrypted with the same key from another file. If each file has a different per-file random portion of its AAD, then AES-GCM encrypted pages from other files can't be swapped in.

Reply:
Yep, lets start this discussion from the scratch. Lets also presume for a second that we don't have the aad_metadata field (which seems to be a source of confusion). 
1. In our storage, we have two encrypted Parquet files, "employees_10_2017" and "employees_May_2018"
2. An attacker breaches our storage and runs `cp employees_10_2017 employees_10_2018`
3. We need to run a query on the latest version of employee table - which we know is written at the end of each month, therefore we know it should be the Oct 2018 version.
4. Without AADs, the attack makes us run query on a wrong, year old table version.
5. But if we use a string "employees_MON_YEAR" as the AAD, then the attack fails. 
6. The writer knows the AAD is "employees_OCT_2018", because it writes the table on the day of 31 Oct 2018.
7. The reader knows the AAD is "employees_OCT_2018", because it runs on the day of 9 Nov 2018, and knows the most recent table was written on 31 Oct 2018.
8. We don't need "aad_metadata" for any of these. Moreover, aad_metadata is unprotected in the file, therefore putting "employees_OCT_2018" there won't help, as this can easily be modified by an attacker.
9. "aad_metadata" has some uses (mentioned in the text), but I'm starting to think maybe they don't justify the cost of confusion caused by this field. We'll either drop this field entirely, or write a very clear and detailed section on its purpose.

Reply:
The example above describes the "smart user" scenario (because both writer and reader know the AAD without a help from any service. They just use a convention "well known" in the organization for table AAD creation). Here goes the "dumb user" scenario - which also doesn't require the aad_metadata field.
5. We use a random string as the file AAD.
6. Writer generates a random string and uses it as an AAD for the file. It also sends the table  name ("employees"),  month of creation, "10_\<CURRENT_YEAR\>", and the generated AAD, to a trusted AAD service. It might also send the file name ("employees_10_\<CURRENT_YEAR\>).
7. Reader runs on the day of 9 Nov 2018 - it goes to the trusted AAD service, and asks for the AAD for the latest version of employees table. The service sends the AAD, and the file name. The reader opens the file, using the AAD to verify its authenticity.
8. As before, we don't need "aad_metadata" for any of these. Putting the random AAD in this field or elsewhere in the file is not safe, can be modified by attacker.

Reply:
The attack I'm addressing is slightly different. In this scenario, users don't know about supplying an AAD. User AAD metadata is empty in the 2017 version and the 2018 version.

The suffix part of the AAD is used to prevent pages from being swapped, so I can't overwrite row group 0 column 1 page 3 with row group 0 column 1 page 4.

But there isn't anything that prevents an attacker from replacing row group 0 column 1 page 4 with the same page from the 2017 version of the file. A per-file unique that is stored in the file would prevent page swapping between any 2 files, even if the unique string is in the file metadata.

This would not protect against copying 2017 over 2018 like your example, but it does at least ensure the pages a file has not been tampered with, in relation to each other.

Reply:
I find the described scenarios a bit oversimplied. Of course, one can use Parquet in the way desribed (separate Parquet files for each month each being read explicitly), but I think in the much more common scenario a directory of Parquet files will correspond to a table without a one-to-one mapping between files and meaningful attributes like months.

When one uses MapReduce or RDD to process big data, different parts of the dataset get processed and written by different machines of the cluster. Instead of writing the data to a single file, each machine writes to a separate file. The individual file names don't even matter.

Additionally, when more data is written to the same table, it is not appended to existing files but new files get added. Impala can't even write arbitrary amount of data into a single Parquet file, it starts a new one for each block instead.

Could the discussed integrity checking measures be used to protect against (partial or full) file-swapping/duplicating/deleting attacks in such a scenario?

Reply:
Yep, the PARQUET-1457 addresses this scenario. As for the AAD prefix section in this doc - I'm currently updating it with more details - with the same focus on integrity of individual files, since Parquet API (ParquetFileWriter and ParquetFileReader)  is designed for writing/reading individual files. Data sets (tables, dataFrames, etc) are handled by frameworks above Parquet. The data set integrity can be handled by framework tools, with or without PARQUET-1457 help. I'll mention this in the doc.